driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/preadv.bpf.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2023 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+#include <helpers/interfaces/fixed_size_event.h>
+#include <helpers/interfaces/variable_size_event.h>
+
+/*=============================== ENTER EVENT ===========================*/
+
+SEC("tp_btf/sys_enter")
+int BPF_PROG(preadv_e,
+	     struct pt_regs *regs,
+	     long id)
+{
+	struct ringbuf_struct ringbuf;
+	if(!ringbuf__reserve_space(&ringbuf, PREADV_E_SIZE))
+	{
+		return 0;
+	}
+
+	ringbuf__store_event_header(&ringbuf, PPME_SYSCALL_PREADV_E);
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	/* Parameter 1: fd (type: PT_FD) */
+	s32 fd = (s32)extract__syscall_argument(regs, 0);
+	ringbuf__store_s64(&ringbuf, (s64)fd);
+
+	/* Parameter 2: pos (type: PT_UINT64) */
+	u64 pos = (u64)extract__syscall_argument(regs, 3);
+	ringbuf__store_u64(&ringbuf, pos);
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	ringbuf__submit_event(&ringbuf);
+
+	return 0;
+}
+
+/*=============================== ENTER EVENT ===========================*/
+
+/*=============================== EXIT EVENT ===========================*/
+
+SEC("tp_btf/sys_exit")
+int BPF_PROG(preadv_x,
+	     struct pt_regs *regs,
+	     long ret)
+{
+	struct auxiliary_map *auxmap = auxmap__get();
+	if(!auxmap)
+	{
+		return 0;
+	}
+
+	auxmap__preload_event_header(auxmap, PPME_SYSCALL_PREADV_X);
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	/* Parameter 1: res (type: PT_ERRNO) */
+	auxmap__store_s64_param(auxmap, ret);
+
+	if(ret > 0)
+	{
+		/* Parameter 2: size (type: PT_UINT32) */
+        auxmap__store_u32_param(auxmap, (u32)ret);
+
+		/* We read the minimum between `snaplen` and what we really
+		 * have in the buffer.
+		 */
+		unsigned long bytes_to_read = maps__get_snaplen();
+
+		if(bytes_to_read > ret)
+		{
+			bytes_to_read = ret;
+		}
```

Comment:
@loresuso @hbrueckner and @Andreagit97 have a question I was holding back when we did the refactor wrt to introducing some upper bounds when applicable. Are there any risks associated with just trusting `ret`?

@loresuso please note that this comment is not related to this PR and rather a general question.

Reply:
> @loresuso @hbrueckner and @Andreagit97 have a question I was holding back when we did the refactor wrt to introducing some upper bounds when applicable. Are there any risks associated with just trusting `ret`?

`ret` typically comes from the kernel implementation of the system call. Not quite sure if you can modify it (before exit) with e.g. `kprobes` or `ptrace

Reply:
This is an interesting question! Do not know of any ways to do that in general, but I do agree with @hbrueckner that this is something that comes from the kernel and we can trust. About modification, I guess it is possible somehow, but doing so would mean already having high privileges on the system for sure, so I wouldn't worry too much

Reply:
Thanks @loresuso and @hbrueckner! Lastly we also never expect the size to be super large right and some safeguard upper bound is not needed here?

Reply:
That check should already be in place, otherwise the verifier could go crazy :) We are going to use `push__bytebuf` which in turn uses `bpf_probe_read_{kernel,user}` and `SAFE_ACCESS` macro to limit which implements the upper bound to `MAX_PARAM_SIZE`

Reply:
agree with @loresuso and @hbrueckner here, `ret` comes directly from the kernel if someone alters it, it means that the kernel is already compromised so probably not an issue